### PR TITLE
Nested Patterns in MExpr

### PR DESCRIPTION
**Pull request by elegios**
This adds nested patterns to the OCaml interpreter and the `mexpr.mc` interpreter. Notably absent is support for nested patterns in `sem` declarations in MLang, but those have some remaining semantic questions (ordering of patterns in composition) and so won't be a part of this PR.

I also haven't yet finished updating `mcore_parser.mc` and `meta.mc`, but I wanted to get the rest looked at sooner rather than later.

As a usability note, it feels surprising when `match e with Con then ...` highlights `Con` as a constructor/type, but the semantics are equivalent with `let Con = e in ...`. It's also annoying to write `None ()` everywhere, but that's expected. As a sidenote, I believe the current implementation allows us to write
```
con None : () -> Option
let None = None ()

let map = lam f. lam o.
  match o with Some x then Some (f x) else
  match o with None _ then None else error "impossible"
```

Note that while we can't get away from using `None _` as the pattern, we can still use `None` as a value.

This gets somewhat close to the inconsistency noted in `mexpr.mc`:

```
con Foo in
let Bar = Foo in
match Foo () with Bar _ then true else false
```

This evaluates to `true` in `mexpr.ml`, but I believe it evaluates to `false` in `mexpr.mc` (both before and after this pull request) (I haven't quite had the time to construct the ast to test it yet).

### CODE REVIEW

**Comment by EliasC** [comment]
> it feels surprising when `match e with Con then ...` highlights `Con` as a constructor/type.

It doesn't exactly help to avoid that error... I will have a look later to see if we can adress that in the Emacs mode.

> `let None = None ()`

I would consider this an anti-pattern.

> [regarding the] the inconsistency noted in `mexpr.mc`

I feel like that program shouldn't compile at all, which lets us avoid making the choice :)

File: src/boot/ast.ml [hunk 1]
```diff
@@ -138,10 +138,18 @@ and tm =
...
 | TmUtest   of info * tm * tm * tm                                  (* Unit testing *)
 
+(* Patterns *)
+and pat =
+| PatNamed of info * ustring                      (* Named, capturing wildcard *)
+| PatTuple of info * pat list                     (* Tuple pattern *)
+| PatCon   of info * ustring * int * pat          (* Constructor pattern *)
+| PatInt   of info * int                          (* Int pattern *)
+| PatBool  of info * bool                         (* Boolean pattern *)
+| PatUnit  of info                                (* Unit pattern *)
```
Comment (EliasC) [inline src/boot/ast.ml:153]: Should this also include character literals?

File: src/boot/ast.ml [hunk 2]
```diff
@@ -173,9 +181,15 @@ module Option = struct
   let map f = function
     | Some x -> Some (f x)
     | None -> None
+  let bind f = function
+    | Some x -> f x
+    | None -> None
+  let value o ~default = match o with
+    | Some x -> x
+    | None -> default
```
Comment (EliasC) [inline src/boot/ast.ml]: Is `value` ever used?

File: src/boot/mexpr.ml [hunk 3]
```diff
@@ -393,7 +393,20 @@ let rec debruijn env t =
   let rec find fi env n x =
     (match env with
      | VarTm(y)::ee -> if y =. x then n else find fi ee (n+1) x
-     | [] -> raise_error fi ("Unknown variable '" ^ Ustring.to_utf8 x ^ "'"))
+     | [] -> raise_error fi ("Unknown variable '" ^ Ustring.to_utf8 x ^ "'")) in
+  let rec dbPat env = function
+    | PatNamed(fi,x) -> (VarTm(x)::env, PatNamed(fi,x))
+    | PatTuple(fi,ps) -> (* NOTE: this causes patterns to introduce names right-to-left, which will cause errors if a later pattern binds a name that is seen as a constructor in a pattern to the left *)
```
Comment (EliasC) [inline src/boot/mexpr.ml:399]: This should also be solved by type checking, right?

File: stdlib/char.mc [hunk 4]
```diff
@@ -1,3 +1,5 @@
+include "seq.mc"
```
Comment (EliasC) [inline stdlib/char.mc:1]: Why is this needed?

File: test/mexpr/nestedpatterns.mc [hunk 5]
```diff
@@ -0,0 +1,25 @@
+mexpr
+
+let classify = lam x.
+  match x with (true, true) then "one" else
+  match x with (true, false) then "two" else
+  match x with (false, true) then "three" else
+  match x with (false, false) then "four"
+  else "five" in
```
Comment (EliasC) [inline test/mexpr/nestedpatterns.mc:8]: Your interpretation of binary numbers is flawed ;)

File: test/mlang/mexpr.mc [hunk 6]
```diff
@@ -216,25 +235,36 @@ end
...
 end
 
+lang BoolPat = Bool
+  syn Pat =
+  | BoolPat Bool
+
+  sem tryMatch (t : Expr) =
+  | BoolPat b ->
+    match t with TmConst CBool b2 then
+      if or (and b b2) (and (not b) (not b2)) then Some [] else None () -- TODO: is there a nicer way to do equality on bools? 'eqb' is unbound
```
Comment (EliasC) [inline test/mlang/mexpr.mc:271]: We should probably add an `eqb`!

**Review by EliasC — APPROVED** [review]
I think this looks good! Do you want this merged as is (after rebasing and possible fixes) or are you planning on adding any features?

File: src/boot/ast.ml (hunk 1 above)
Comment (elegios) [reply src/boot/ast.ml:153]: Definitely

File: src/boot/ast.ml (hunk 2 above)
Comment (elegios) [reply src/boot/ast.ml]: It used to be, but no longer, I'll remove it

File: stdlib/char.mc (hunk 4 above)
Comment (elegios) [reply stdlib/char.mc:1]: A use of `head`. It works when `char.mc` is included somewhere else, since `prelude` is already included at that point, but when running tests on the standard library I run them without `prelude`, since it would induce an import cycle otherwise (well, only on the files that are included in `prelude`, but still, that's annoying to detect for the test runner).

File: test/mexpr/nestedpatterns.mc (hunk 5 above)
Comment (elegios) [reply test/mexpr/nestedpatterns.mc:8]: :)

File: src/boot/mexpr.ml (hunk 3 above)
Comment (elegios) [reply src/boot/mexpr.ml:399]: Mostly yes, but it will at least be solved by #31.

It does still leave shadowing for normal names though, so 

```
match e with (a, a) then a else ...
```

will use the left value, since it shadows the right. This may or may not be expected.

Comment (EliasC) [reply src/boot/mexpr.ml:399]: My feeling is that shadowing in the same pattern is also something that a typechecker should prevent.
